src/main/java/org/springframework/integration/dsl/StandardIntegrationFlow.java

```diff
@@ -88,7 +85,9 @@ public void start() {
 	public void stop(Runnable callback) {
 		if (this.running) {
 			AggregatingCallback aggregatingCallback = new AggregatingCallback(this.lifecycles.size(), callback);
-			for (SmartLifecycle lifecycle : this.lifecycles) {
+			ListIterator<SmartLifecycle> iterator = this.lifecycles.listIterator(this.lifecycles.size());
+			while (iterator.hasPrevious()) {
```

Comment:
Correct.
That's why I do that reverse complex logic in the `start()`.
I iterate `this.integrationComponents` from right -> left and populate `lifecycles` as is.
Therefore the first item in the `lifecycles` is that which has to be stopped as last.
That's why a reverse iteration is here in the `stop()`.

I won't mind if you suggest more simpler logic.

Thanks.